models/classes/TestSessionState/TestSessionStateRestorationService.php

```diff
@@ -0,0 +1,186 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2022 (original work) Open Assessment Technologies SA
+ *
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoQtiTest\models\TestSessionState;
+
+use common_exception_NoContent;
+use common_exception_NotFound;
+use League\Flysystem\FileNotFoundException;
+use oat\tao\model\state\StateMigration;
+use oat\taoDelivery\model\execution\DeliveryExecution;
+use oat\taoQtiTest\models\QtiTestExtractionFailedException;
+use oat\taoQtiTest\models\QtiTestUtils;
+use oat\taoQtiTest\models\runner\ExtendedState;
+use oat\taoQtiTest\models\runner\time\QtiTimeStorage;
+use oat\taoQtiTest\models\TestSessionService;
+use oat\taoQtiTest\models\TestSessionState\Api\TestSessionStateRestorationInterface;
+use oat\taoQtiTest\models\TestSessionState\Exception\RestorationImpossibleException;
+use Psr\Log\LoggerInterface;
+use qtism\data\AssessmentSection;
+use qtism\data\SectionPartCollection;
+use qtism\data\TestPartCollection;
+
+class TestSessionStateRestorationService implements TestSessionStateRestorationInterface
+{
+    /** @var TestSessionService */
+    private $testSessionService;
+    /** @var QtiTestUtils */
+    private $qtiTestUtils;
+    /** @var StateMigration */
+    private $stateMigration;
+    /** @var LoggerInterface */
+    private $logger;
+
+    public function __construct(
+        TestSessionService $testSessionService,
+        QtiTestUtils $qtiTestUtils,
+        StateMigration $stateMigration,
+        LoggerInterface $logger
+    ) {
+        $this->testSessionService = $testSessionService;
+        $this->qtiTestUtils = $qtiTestUtils;
+        $this->stateMigration = $stateMigration;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function restore(DeliveryExecution $deliveryExecution): void
+    {
+        $deliveryExecutionId = $deliveryExecution->getIdentifier();
+        $userId = $deliveryExecution->getUserIdentifier();
+
+        $this->restoreTestSessionState($userId, $deliveryExecutionId);
+        $this->restoreExtendedState($userId, $deliveryExecutionId);
+        $this->restoreTimeLineState($userId, $deliveryExecutionId);
+        $this->restoreItemsState($deliveryExecution);
```

Comment:
Once the restoration has happened, and the archiving is getting requested a second time, it will fail with `File already exists at path: stateBackup/<id>`.

I guess it should be safe to remove the state after all the state levels have been restored.

WDYT of adding $this->stateMigration->removeBackup() calls on each restored piece of data after _all of them_ have been restored?

Reply:
Nice catch. Didn't think about that part. Probably will be good to trigger async tasks for that